vvm/utils/versioning.py

```diff
@@ -0,0 +1,105 @@
+import itertools
+import re
+from typing import Optional
+
+from packaging.specifiers import InvalidSpecifier, Specifier
+from packaging.version import Version
+
+from vvm.exceptions import UnexpectedVersionError
+from vvm.install import get_installable_vyper_versions, get_installed_vyper_versions
+
+_VERSION_RE = re.compile(r"\s*#\s*(?:pragma\s+|@)version\s+([=><^~]*)(\d+\.\d+\.\d+\S*)")
+
+# this is used to convert from the npm-style (before 0.4) to the pypi-style (0.4+)
+_SPECIFIER_OVERRIDES = {"^": "~=", "~": ">="}
+
+
+def _detect_version_specifier(source_code: str) -> Optional[Specifier]:
+    """
+    Detect the version given by the pragma version in the source code.
+
+    Arguments
+    ---------
+    source_code : str
+        Source code to detect the version from.
+
+    Returns
+    -------
+    str
+        vyper version specifier, or None if none could be detected.
+    """
+    try:
+        match = next(_VERSION_RE.finditer(source_code))
+        specifier, version_str = match.groups()
+        if specifier in ("~", "^"):
+            # convert from npm-style to pypi-style
+            if specifier == "^":
+                # minor match, remove the patch from the version
+                version_str = ".".join(version_str.split(".")[:-1])
+            specifier = "~="  # finds compatible versions
+        specifier = _SPECIFIER_OVERRIDES.get(specifier, specifier) or "=="
+        return Specifier(specifier + version_str)
+    except (StopIteration, InvalidSpecifier):
+        return None
```

Comment:
also with no try/catch block, hard to follow at first glance where the exceptional cases happen
```suggestion
    match = _VERSION_RE.search(source_code)
    if match is None:
        return None

    specifier, version_str = match.groups()
    if specifier in ("~", "^"):
        # convert from npm-style to pypi-style
        if specifier == "^":
            # minor match, remove the patch from the version
            version_str = ".".join(version_str.split(".")[:-1])
        specifier = "~="  # finds compatible versions
    specifier = _SPECIFIER_OVERRIDES.get(specifier, specifier) or "=="
    try:
        return Specifier(specifier + version_str)
    except InvalidSpecifier:
        return None
```

Reply:
Agree with you. It used to make sense when the method was just 2 lines